lib/Arch/X86_64_SysV.cpp

```diff
@@ -0,0 +1,158 @@
+      auto struct_ptr = llvm::cast<llvm::StructType>(value_declaration.type);
+      auto mapping =
+          TryReturnThroughRegisters(*struct_ptr, return_register_constraints);
+      if (mapping) {
+        // There is a valid split over registers, so add the mapping
+        return *mapping;
+      } else {
+        // Struct splitting didn't work so do RVO. Assume that the pointer
+        // to the return value resides in RAX.
+        injected_sret = true;
+        value_declaration.reg = arch->RegisterByName("RAX");
+      }
+      break;
+    }
+    default: {
+      LOG(ERROR) << "Encountered an unknown return type";
+      exit(1);
+    }
+  }
+
+  return_value_declarations.push_back(value_declaration);
+
+  return return_value_declarations;
+}
+
+// For X86_64_SysV, the general argument passing behavior is, try to pass the
+// arguments in registers RDI, RSI, RDX, RCX, R8, R9 from integral types and
+// XMM0 - XMM7 for float types. If the argument is a struct but can be
+// completely split over the above registers, then greedily split it over the
+// registers. Otherwise, the struct is passed entirely on the stack. If we run
+// our of registers then pass the rest of the arguments on the stack.
+std::vector<anvill::ParameterDecl> X86_64_SysV::BindParameters(
+    const llvm::Function &function, bool injected_sret) {
+  std::vector<anvill::ParameterDecl> parameter_declarations;
+  auto param_names = TryRecoverParamNames(function);
+  llvm::DataLayout dl(function.getParent());
+
+  // Used to keep track of which registers have been allocated
+  std::vector<bool> allocated(parameter_register_constraints.size(), false);
+
+  // Stack offset describes the stack position of the first stack argument on
+  // entry to the callee. For X86_64_SysV, this is [rsp + 8] since there is the
+  // return address at [rsp].
+  unsigned int stack_offset = 8;
+
+  // If there is an injected sret (an implicit sret) then we need to allocate
+  // the first parameter to the sret struct. The type of said sret parameter
+  // will be the return type of the function.
+  if (injected_sret) {
+    anvill::ParameterDecl decl = {};
+    decl.name = "param0";
+    decl.type = function.getReturnType();
+    decl.reg = arch->RegisterByName("RAX");
+    allocated[0] = true;
+    parameter_declarations.push_back(decl);
+  }
+
+  for (auto &argument : function.args()) {
+    anvill::ParameterDecl declaration = {};
+    declaration.type = argument.getType();
+
+    // Try to allocate from a register. If a register is not available then
+    // allocate from the stack.
+    if (const remill::Register *reg = TryRegisterAllocate(
+            *argument.getType(), allocated, parameter_register_constraints)) {
+      declaration.reg = reg;
+    } else {
+      declaration.mem_offset = stack_offset;
+      declaration.mem_reg = arch->RegisterByName("RSP");
+      stack_offset += dl.getTypeAllocSize(argument.getType());
```

Comment:
I think we should see if alignment changes anything. For example, can we arrange for a stack-allocated parameter of type `struct alignas(N) { char x; };` with `N` like `1`, `3`, `7`, `11`, `15`, etc. to see if more space is allocated than just what LLVM uses to represent the type.

Reply:
I am pretty sure I am using getTypeAllocSize like we talked about, so I will leave this for now.